### PR TITLE
Refactor runner - Remove 'DoneWhenAllIdle', Add 'ActiveThreads' counter

### PR DESCRIPTION
**Pull request by omerzi**
- [x] All [tests](https://github.com/jfrog/gofrog#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
- [x] I labeled this pull request with one of the following: 'breaking change', 'new feature', 'bug', or 'ignore for release'

---

### CODE REVIEW

File: parallel/runner.go [hunk 1]
```diff
@@ -44,20 +42,18 @@ type runner struct {
 	maxParallel int
 	// If true, the runner will be cancelled on the first error thrown from a task.
 	failFast bool
+	// Alerts that the runner received some tasks and started running them.
```
Comment (eyalbe4) [inline parallel/runner.go]: Alerts --> Indicates
running --> executing

File: parallel/runner.go [hunk 2]
```diff
@@ -129,11 +125,16 @@ func (r *runner) Run() {
 	r.threadsWaitGroup.Wait()
 }
 
-// The producer notifies that no more tasks will be produced.
+// Done notifies that no more tasks will be produced.
```
Comment (eyalbe4) [inline parallel/runner.go]: Done notifies... --> Done is used to notify...

File: parallel/runner.go [hunk 3]
```diff
@@ -237,17 +185,22 @@ func (r *runner) addThread() {
...
+		r.openThreads++
+		r.openThreadsLock.Unlock()
 
 		// Keep on taking tasks from the queue.
 		for t := range r.tasks {
-			r.idle.Store(threadId, false)
-
+			// Increase num of threads running tasks
+			atomic.AddUint32(&r.activeThreads, 1)
+			// Mark that some task was executed and the runner started with tasks executing.
```
Comment (eyalbe4) [inline parallel/runner.go]: Not sure the phrasing of thsi comment is the best. Anyway, I think a description for the below code is not needed.

File: parallel/runner.go [hunk 4]
```diff
@@ -237,17 +185,22 @@ func (r *runner) addThread() {
...
+			if !r.started {
```
Comment (eyalbe4) [inline parallel/runner.go]: Is this if statement really needed?
Can't we just set `r.started = true` without checking the value first?

File: parallel/runner.go [hunk 5]
```diff
@@ -237,17 +185,22 @@ func (r *runner) addThread() {
...
-		r.runningThreadsLock.Unlock()
+		r.openThreadsLock.Lock()
+		r.openThreads++
+		r.openThreadsLock.Unlock()
 
 		// Keep on taking tasks from the queue.
 		for t := range r.tasks {
-			r.idle.Store(threadId, false)
-
+			// Increase num of threads running tasks
```
Comment (eyalbe4) [inline parallel/runner.go]: // Increase num of threads running tasks
-->
// Increase the total of active threads.

File: parallel/runner.go [hunk 6]
```diff
@@ -237,17 +185,22 @@ func (r *runner) addThread() {
...
+			atomic.AddUint32(&r.activeThreads, 1)
+			// Mark that some task was executed and the runner started with tasks executing.
+			if !r.started {
+				r.started = true
+			}
 			// Run the task.
 			e := t.run(threadId)
+			// Decrease num of threads running tasks
```
Comment (eyalbe4) [inline parallel/runner.go]: // Decrease num of threads running tasks
-->
// Decrease the total of active threads.

File: parallel/runner.go [hunk 7]
```diff
@@ -263,20 +216,15 @@ func (r *runner) addThread() {
 					break
 				}
 			}
-			r.idle.Store(threadId, true)
-			// Save the current time as the thread's last active time.
-			r.lastActive.Store(threadId, strconv.FormatInt(time.Now().Unix(), 10))
 
-			r.runningThreadsLock.Lock()
+			r.openThreadsLock.Lock()
 			// If there are more open threads than maxParallel, then this thread will be closed.
```
Comment (eyalbe4) [inline parallel/runner.go]: // If there are more open threads than maxParallel, then this thread will be closed.
-->
// If there the total of open threads is larger than the maximim (maxParallel), then this thread should be closed.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: parallel/runner.go (hunk 2 above)
Comment (omerzi) [reply parallel/runner.go]: done

File: parallel/runner.go (hunk 4 above)
Comment (omerzi) [reply parallel/runner.go]: done

File: parallel/runner.go (hunk 3 above)
Comment (omerzi) [reply parallel/runner.go]: done

File: parallel/runner.go (hunk 1 above)
Comment (omerzi) [reply parallel/runner.go]: done

File: parallel/runner.go (hunk 5 above)
Comment (omerzi) [reply parallel/runner.go]: done

File: parallel/runner.go (hunk 6 above)
Comment (omerzi) [reply parallel/runner.go]: done

File: parallel/runner.go (hunk 7 above)
Comment (omerzi) [reply parallel/runner.go]: done

File: parallel/runner.go [hunk 8]
```diff
@@ -237,17 +193,21 @@ func (r *runner) addThread() {
 	nextThreadId := atomic.AddUint32(&r.threadCount, 1) - 1
 	go func(threadId int) {
 		defer r.threadsWaitGroup.Done()
-
-		r.runningThreadsLock.Lock()
-		r.runningThreads++
-		r.runningThreadsLock.Unlock()
+		r.openThreadsLock.Lock()
+		r.openThreads++
```
Comment (eyalbe4) [inline parallel/runner.go:197]: I suggest we use an atomic int here, and by doing so, remove `openThreadsLock ` altogether.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: parallel/runner.go (hunk 8 above)
Comment (omerzi) [reply parallel/runner.go:197]: I can use atomic int here, but I think that I still need the lock because of the if statement in line 229

**Pull request merged**